src/OpenTelemetry/Trace/Configuration/OpenTelemetrySdk.cs

```diff
@@ -72,66 +72,47 @@ public static IDisposable EnableOpenTelemetry(Action<OpenTelemetryBuilder> confi
-                // This creates Activity and sets its IsAllDataRequested to true.
-                // Library authors can check activity.IsAllDataRequested and avoid
-                // doing any additional telemetry population.
-                // Activity.IsAllDataRequested is the equivalent of Span.IsRecording
-                //
-                // If Sampler returns IsSampled as false, returns ActivityDataRequest.None
-                // This prevents Activity from being created at all.
-                GetRequestedDataUsingContext = (ref ActivityCreationOptions<ActivityContext> options) =>
-                {
-                    BuildSamplingParameters(options, out var samplingParameters);
-                    var shouldSample = sampler.ShouldSample(samplingParameters);
-                    if (shouldSample.IsSampled)
-                    {
-                        return ActivityDataRequest.AllDataAndRecorded;
-                    }
-                    else
-                    {
-                        return ActivityDataRequest.None;
-                    }
-
-                    // TODO: Improve this to properly use ActivityDataRequest.AllData, PropagationData as well.
-                },
+                // This delegate informs ActivitySource about sampling decision when the parent context is an ActivityContext.
+                GetRequestedDataUsingContext = (ref ActivityCreationOptions<ActivityContext> options) => ComputeActivityDataRequest(options, sampler),
             };
 
             ActivitySource.AddActivityListener(listener);
 
             return listener;
         }
 
-        internal static void BuildSamplingParameters(
-            in ActivityCreationOptions<ActivityContext> options, out ActivitySamplingParameters samplingParameters)
+        internal static ActivityDataRequest ComputeActivityDataRequest(
+            in ActivityCreationOptions<ActivityContext> options,
+            ActivitySampler sampler)
         {
-            ActivityContext parentContext = options.Parent;
-            if (parentContext == default)
-            {
-                // Check if there is already a parent for the current activity.
-                var parentActivity = Activity.Current;
-                if (parentActivity != null)
-                {
-                    parentContext = parentActivity.Context;
-                }
-            }
+            var isRootSpan = options.Parent.TraceId == default;
 
             // This is not going to be the final traceId of the Activity (if one is created), however, it is
             // needed in order for the sampling to work. This differs from other OTel SDKs in which it is
             // the Sampler always receives the actual traceId of a root span/activity.
-            ActivityTraceId traceId = parentContext.TraceId != default
-                ? parentContext.TraceId
+            ActivityTraceId traceId = !isRootSpan
+                ? options.Parent.TraceId
```

Comment:
do you mean automatically setting the parent trace id? if so and I understand correctly, I think this handling shouldn't be in BCL as the API will be kind of not ne honest reporting the data to the caller. the callers will not know if we generated some id for them or we used the parent id. I think it make sense to keep handling this here.

---
In reply to: [438443909](https://github.com/open-telemetry/opentelemetry-dotnet/pull/719#discussion_r438443909) [](ancestors = 438443909)

Reply:
Correct @tarekgh, and at this point I agree with your statement. Now we need to make a way for the trace ID generated on the callback to be used in the Activity creation. My first thought was to make `options.Parent` settable, this way the callback has full control of what to do in this case. A related benefit specific to OTel is that the sampler can directly receive `options`, with the caveat that the `traceId` is now in the parent context.

Reply:
The concern with making `Parent` settable is what we'll do when having multiple listeners and the different listeners can set the parent to a different value? would it make sense to have the sampler handle the default trace id instead? the sampler will have the full context at that time to decide I think. 

CC @noahfalk

Reply:
Right now it is not possible for the sampler or OTel SDK to do that. The sampler is called before the actual trace ID is created. What about a flag/option to create the trace ID for the root span? I mean at https://github.com/tarekgh/runtime/blob/dc0f9e082a3383d21b4c94b116f5daea285f98fb/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L136 replace the last `context` with a context with a new trace ID, something like:
```c#
ActivityContext initializedContext =  context == default && Activity.Current != null ? Activity.Current.Context : ActivityContext.NewRootContext();
```
this will be consistent for all listeners and if behind an option configuration the price is paid only when required. Thoughts?

Reply:
You know how the sampler works better than me but I am wondering about how the sampler is using the trace Id? I mean if the parent trace id = default and we pass that to the sampler, can't the sampler decide using this information? think about it as default Id is some constant trace Id which the smaller can use.

Reply:
Specifically for the probabilistic sampler, the trace ID is the "random input" used to make the sampling decision. The trace ID is always non-zero when passed to the sampler, if it is a root span a new trace ID is created before the sampling decision is known. So the probabilistic sampler needs a valid trace ID to make its decision.

Reply:
@pjanotti thanks for the clarification.  

> Specifically for the probabilistic sampler, the trace ID is the "random input" used to make the sampling decision

Does it mean passing any random trace Id to the tracer will make it work? if decided to create Activity after this sampling decision, does it matter if the Activity created with this trace Id or not?

Reply:
> Does it mean passing any random trace Id to the tracer will make it work? if decided to create Activity after this sampling decision, does it matter if the Activity created with this trace Id or not?

Yes, it matters if it is different. Right now the probabilistic sampler is working with a "local trace ID" not the actual trace ID, it works because for the probabilistic sampler it respects the decision for the root span (independent of the trace ID). However, in the way that the spec is written, one could certainly create a sampler that checks for bits of the trace ID to make its decision and in this case, the different trace IDs will be a problem.

Reply:
OTel also wants to specify how the probabilistic sample should treat the trace ID across different languages. In this case, an outgoing context from .NET may have a trace ID that shouldn't be sampled but was flagged as sampled because of the "local trace ID" so it may look like a bug compared to the other languages.

Reply:
For a rootspan:
ProbabilitySampler gets traceid.generateRandom = traceid_random1, and makes decision not sample.
We return PropagationData, hence root Activity will be generated by with traceid_random2.

Now a child activity is being created. 
ProbabilitySampler gets traceid_random2. Since parent was not sampled, sampling algorithm now runs with traceid_random2. It can chose to sample in child. 

At this stage - we already have broken trace right? As the child is sampled in, and parent is not?

Reply:
Per spec, the probability sampler default behavior should be to trust the parent span however it is allowed to be configurable and depending on the configuration it can definitely break traces due to the difference in trace ID.

Reply:
@tarekgh and I discussed this offline a bit more - the understanding is .NET Activity will expose some mechanism to handle this.
One idea is to make an additional parameter "ProposedActivityTraceId" as part of ActivityCreationoptions. If parent is invalid, then sampler can make decision based on "ProposedActivityTraceId", and if the Activity does end up getting created, it'll be using this "ProposedActivityTraceId"

Or alternately, put the proposedTraceId in the ActivityContext itself - if there parent is valid, then ActivityContext contains its TraceId, SpanId. If no parent, then TraceID will contain proposedTraceId, and OTSDK can determine that this is the proposedID, by seeing that spanid is default.

(Just sharing what we discussed - I'll let .NET team to propose this, and we can discuss more)

Happy to merge this PR now.

Reply:
Yes, I'll add this to the .NET changes proposal. I have one question though, what the sampler is going to do in the GetRequestedDataUsingParentId? this one, there is no context/trace id is passed and will only have the parent Id.